{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Python app: Find proteins containing c-terminal degrons\n",
    "\n",
    "This app finds and selects all proteins that contain c-terminal degrons."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. Upload human gene data &rarr; anotate gene_name, transcript_name, protein_name\n",
    "\n",
    "2. Transcribe and translate &rarr; anotate proteins\n",
    "\n",
    "3. Find c-degron sequences: use consensus sequences  \n",
    "\n",
    "|C-degrons|\n",
    "|---------:|\n",
    "|-GG|  \n",
    "|-RG|  \n",
    "|-PG|  \n",
    "|-XR|  \n",
    "|-RXXG|  \n",
    "|-EE| \n",
    "|-RXX|  \n",
    "|-VX|  \n",
    "|-AX|  \n",
    "|-A|    \n",
    "\n",
    "Varshavsky *et al* 2019 (**Fig S3** Supl material)  \n",
    "Lin *et al* 2018  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1. Upload genomic data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on class EnsemblRelease in module pyensembl.ensembl_release:\n",
      "\n",
      "class EnsemblRelease(pyensembl.genome.Genome)\n",
      " |  EnsemblRelease(release=99, species=Species(latin_name='homo_sapiens', synonyms=['human'], reference_assemblies={'GRCh38': (76, 99), 'GRCh37': (55, 75), 'NCBI36': (54, 54)}), server='ftp://ftp.ensembl.org')\n",
      " |  \n",
      " |  Bundles together the genomic annotation and sequence data associated with\n",
      " |  a particular release of the Ensembl database.\n",
      " |  \n",
      " |  Method resolution order:\n",
      " |      EnsemblRelease\n",
      " |      pyensembl.genome.Genome\n",
      " |      serializable.serializable.Serializable\n",
      " |      builtins.object\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __eq__(self, other)\n",
      " |      Return self==value.\n",
      " |  \n",
      " |  __hash__(self)\n",
      " |      Return hash(self).\n",
      " |  \n",
      " |  __init__(self, release=99, species=Species(latin_name='homo_sapiens', synonyms=['human'], reference_assemblies={'GRCh38': (76, 99), 'GRCh37': (55, 75), 'NCBI36': (54, 54)}), server='ftp://ftp.ensembl.org')\n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      reference_name : str\n",
      " |          Name of genome assembly which annotations in GTF are aligned against\n",
      " |          (and from which sequence data is drawn)\n",
      " |      \n",
      " |      annotation_name : str\n",
      " |          Name of annotation source (e.g. \"Ensembl)\n",
      " |      \n",
      " |      annotation_version : int or str\n",
      " |          Version of annotation database (e.g. 75)\n",
      " |      \n",
      " |      gtf_path_or_url : str\n",
      " |          Path or URL of GTF file\n",
      " |      \n",
      " |      transcript_fasta_paths_or_urls : list\n",
      " |          List of paths or URLs of FASTA files containing transcript sequences\n",
      " |      \n",
      " |      protein_fasta_paths_or_urls : list\n",
      " |          List of paths or URLs of FASTA files containing protein sequences\n",
      " |      \n",
      " |      decompress_on_download : bool\n",
      " |          If remote file is compressed, decompress the local copy?\n",
      " |      \n",
      " |      copy_local_files_to_cache : bool\n",
      " |          If genome data file is local use it directly or copy to cache first?\n",
      " |      \n",
      " |      cache_directory_path : None\n",
      " |          Where to place downloaded and cached files for this genome,\n",
      " |          by default inferred from reference name, annotation name,\n",
      " |          annotation version, and global cache dir for pyensembl.\n",
      " |  \n",
      " |  __str__(self)\n",
      " |      Return str(self).\n",
      " |  \n",
      " |  install_string(self)\n",
      " |      Add every missing file to the install string shown to the user\n",
      " |      in an error message.\n",
      " |  \n",
      " |  to_dict(self)\n",
      " |      Returns a dictionary of the essential fields of this Genome.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Class methods defined here:\n",
      " |  \n",
      " |  cached(release=99, species=Species(latin_name='homo_sapiens', synonyms=['human'], reference_assemblies={'GRCh38': (76, 99), 'GRCh37': (55, 75), 'NCBI36': (54, 54)}), server='ftp://ftp.ensembl.org') from builtins.type\n",
      " |      Construct EnsemblRelease if it's never been made before, otherwise\n",
      " |      return an old instance.\n",
      " |  \n",
      " |  from_dict(state_dict) from builtins.type\n",
      " |      Deserialize EnsemblRelease without creating duplicate instances.\n",
      " |  \n",
      " |  normalize_init_values(release, species, server) from builtins.type\n",
      " |      Normalizes the arguments which uniquely specify an EnsemblRelease\n",
      " |      genome.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data and other attributes defined here:\n",
      " |  \n",
      " |  _genome_cache = <WeakValueDictionary>\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from pyensembl.genome.Genome:\n",
      " |  \n",
      " |  __repr__(self)\n",
      " |      Return repr(self).\n",
      " |  \n",
      " |  _all_feature_values(self, column, feature, distinct=True, contig=None, strand=None)\n",
      " |      Cached lookup of all values for a particular feature property from\n",
      " |      the database, caches repeated queries in memory and\n",
      " |      stores them as a CSV.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      \n",
      " |      column : str\n",
      " |          Name of property (e.g. exon_id)\n",
      " |      \n",
      " |      feature : str\n",
      " |          Type of entry (e.g. exon)\n",
      " |      \n",
      " |      distinct : bool, optional\n",
      " |          Keep only unique values\n",
      " |      \n",
      " |      contig : str, optional\n",
      " |          Restrict query to particular contig\n",
      " |      \n",
      " |      strand : str, optional\n",
      " |          Restrict results to \"+\" or \"-\" strands\n",
      " |      \n",
      " |      Returns a list constructed from query results.\n",
      " |  \n",
      " |  _fields(self)\n",
      " |  \n",
      " |  _get_cached_path(self, field_name, path_or_url, download_if_missing=False, overwrite=False)\n",
      " |      Get the local path for a possibly remote file, invoking either\n",
      " |      a download or install error message if it's missing.\n",
      " |  \n",
      " |  _get_gtf_path(self, download_if_missing=False, overwrite=False)\n",
      " |  \n",
      " |  _get_protein_fasta_paths(self, download_if_missing=False, overwrite=False)\n",
      " |  \n",
      " |  _get_transcript_fasta_paths(self, download_if_missing=False, overwrite=False)\n",
      " |  \n",
      " |  _init_lazy_fields(self)\n",
      " |      Member data that gets loaded or constructed on demand\n",
      " |  \n",
      " |  _query_exon_ids(self, property_name, value)\n",
      " |  \n",
      " |  _query_gene_ids(self, property_name, value, feature='gene')\n",
      " |  \n",
      " |  _query_gene_name(self, property_name, property_value, feature_type)\n",
      " |  \n",
      " |  _query_transcript_ids(self, property_name, value, feature='transcript')\n",
      " |  \n",
      " |  _query_transcript_names(self, property_name, value)\n",
      " |  \n",
      " |  _set_local_paths(self, download_if_missing=False, overwrite=False)\n",
      " |  \n",
      " |  clear_cache(self)\n",
      " |      Clear any in-memory cached values and short-lived on-disk\n",
      " |      materializations from MemoryCache\n",
      " |  \n",
      " |  contigs(self)\n",
      " |      Returns all contig names for any gene in the genome\n",
      " |      (field called \"seqname\" in Ensembl GTF files)\n",
      " |  \n",
      " |  delete_index_files(self)\n",
      " |      Delete all data aside from source GTF and FASTA files\n",
      " |  \n",
      " |  download(self, overwrite=False)\n",
      " |      Download data files needed by this Genome instance.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      overwrite : bool, optional\n",
      " |          Download files regardless whether local copy already exists.\n",
      " |  \n",
      " |  exon_by_id(self, exon_id)\n",
      " |      Construct an Exon object from its ID by looking up the exon\"s\n",
      " |      properties in the given Database.\n",
      " |  \n",
      " |  exon_ids(self, contig=None, strand=None)\n",
      " |  \n",
      " |  exon_ids_at_locus(self, contig, position, end=None, strand=None)\n",
      " |  \n",
      " |  exon_ids_of_gene_id(self, gene_id)\n",
      " |  \n",
      " |  exon_ids_of_gene_name(self, gene_name)\n",
      " |  \n",
      " |  exon_ids_of_transcript_id(self, transcript_id)\n",
      " |  \n",
      " |  exon_ids_of_transcript_name(self, transcript_name)\n",
      " |  \n",
      " |  exons(self, contig=None, strand=None)\n",
      " |      Create exon object for all exons in the database, optionally\n",
      " |      restrict to a particular chromosome using the `contig` argument.\n",
      " |  \n",
      " |  exons_at_locus(self, contig, position, end=None, strand=None)\n",
      " |  \n",
      " |  gene_by_id(self, gene_id)\n",
      " |      Construct a Gene object for the given gene ID.\n",
      " |  \n",
      " |  gene_by_protein_id(self, protein_id)\n",
      " |      Get the gene ID associated with the given protein ID,\n",
      " |      return its Gene object\n",
      " |  \n",
      " |  gene_id_of_protein_id(self, protein_id)\n",
      " |      What is the gene ID associated with a given protein ID?\n",
      " |  \n",
      " |  gene_ids(self, contig=None, strand=None)\n",
      " |      What are all the gene IDs\n",
      " |      (optionally restrict to a given chromosome/contig and/or strand)\n",
      " |  \n",
      " |  gene_ids_at_locus(self, contig, position, end=None, strand=None)\n",
      " |  \n",
      " |  gene_ids_of_gene_name(self, gene_name)\n",
      " |      What are the gene IDs associated with a given gene name?\n",
      " |      (due to copy events, there might be multiple genes per name)\n",
      " |  \n",
      " |  gene_name_of_exon_id(self, exon_id)\n",
      " |  \n",
      " |  gene_name_of_gene_id(self, gene_id)\n",
      " |  \n",
      " |  gene_name_of_transcript_id(self, transcript_id)\n",
      " |  \n",
      " |  gene_name_of_transcript_name(self, transcript_name)\n",
      " |  \n",
      " |  gene_names(self, contig=None, strand=None)\n",
      " |      Return all genes in the database,\n",
      " |      optionally restrict to a chromosome and/or strand.\n",
      " |  \n",
      " |  gene_names_at_locus(self, contig, position, end=None, strand=None)\n",
      " |  \n",
      " |  genes(self, contig=None, strand=None)\n",
      " |      Returns all Gene objects in the database. Can be restricted to a\n",
      " |      particular contig/chromosome and strand by the following arguments:\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      contig : str\n",
      " |          Only return genes on the given contig.\n",
      " |      \n",
      " |      strand : str\n",
      " |          Only return genes on this strand.\n",
      " |  \n",
      " |  genes_at_locus(self, contig, position, end=None, strand=None)\n",
      " |  \n",
      " |  genes_by_name(self, gene_name)\n",
      " |      Get all the unqiue genes with the given name (there might be multiple\n",
      " |      due to copies in the genome), return a list containing a Gene object\n",
      " |      for each distinct ID.\n",
      " |  \n",
      " |  index(self, overwrite=False)\n",
      " |      Assuming that all necessary data for this Genome has been downloaded,\n",
      " |      generate the GTF database and save efficient representation of\n",
      " |      FASTA sequence files.\n",
      " |  \n",
      " |  loci_of_gene_names(self, gene_name)\n",
      " |      Given a gene name returns list of Locus objects with fields:\n",
      " |          chromosome, start, stop, strand\n",
      " |      You can get multiple results since a gene might have multiple copies\n",
      " |      in the genome.\n",
      " |  \n",
      " |  locus_of_exon_id(self, exon_id)\n",
      " |      Given an exon ID returns Locus\n",
      " |  \n",
      " |  locus_of_gene_id(self, gene_id)\n",
      " |      Given a gene ID returns Locus with: chromosome, start, stop, strand\n",
      " |  \n",
      " |  locus_of_transcript_id(self, transcript_id)\n",
      " |  \n",
      " |  protein_ids(self, contig=None, strand=None)\n",
      " |      What are all the protein IDs\n",
      " |      (optionally restrict to a given chromosome and/or strand)\n",
      " |  \n",
      " |  protein_ids_at_locus(self, contig, position, end=None, strand=None)\n",
      " |  \n",
      " |  protein_sequence(self, protein_id)\n",
      " |      Return cDNA nucleotide sequence of transcript, or None if\n",
      " |      transcript doesn't have cDNA sequence.\n",
      " |  \n",
      " |  required_local_files(self)\n",
      " |  \n",
      " |  required_local_files_exist(self, empty_files_ok=False)\n",
      " |  \n",
      " |  transcript_by_id(self, transcript_id)\n",
      " |      Construct Transcript object with given transcript ID\n",
      " |  \n",
      " |  transcript_by_protein_id(self, protein_id)\n",
      " |  \n",
      " |  transcript_id_of_protein_id(self, protein_id)\n",
      " |      What is the transcript ID associated with a given protein ID?\n",
      " |  \n",
      " |  transcript_ids(self, contig=None, strand=None)\n",
      " |  \n",
      " |  transcript_ids_at_locus(self, contig, position, end=None, strand=None)\n",
      " |  \n",
      " |  transcript_ids_of_exon_id(self, exon_id)\n",
      " |  \n",
      " |  transcript_ids_of_gene_id(self, gene_id)\n",
      " |  \n",
      " |  transcript_ids_of_gene_name(self, gene_name)\n",
      " |  \n",
      " |  transcript_ids_of_transcript_name(self, transcript_name)\n",
      " |  \n",
      " |  transcript_name_of_transcript_id(self, transcript_id)\n",
      " |  \n",
      " |  transcript_names(self, contig=None, strand=None)\n",
      " |      What are all the transcript names in the database\n",
      " |      (optionally, restrict to a given chromosome and/or strand)\n",
      " |  \n",
      " |  transcript_names_at_locus(self, contig, position, end=None, strand=None)\n",
      " |  \n",
      " |  transcript_names_of_gene_name(self, gene_name)\n",
      " |  \n",
      " |  transcript_sequence(self, transcript_id)\n",
      " |      Return cDNA nucleotide sequence of transcript, or None if\n",
      " |      transcript doesn't have cDNA sequence.\n",
      " |  \n",
      " |  transcripts(self, contig=None, strand=None)\n",
      " |      Construct Transcript object for every transcript entry in\n",
      " |      the database. Optionally restrict to a particular\n",
      " |      chromosome using the `contig` argument.\n",
      " |  \n",
      " |  transcripts_at_locus(self, contig, position, end=None, strand=None)\n",
      " |  \n",
      " |  transcripts_by_name(self, transcript_name)\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors inherited from pyensembl.genome.Genome:\n",
      " |  \n",
      " |  db\n",
      " |  \n",
      " |  protein_sequences\n",
      " |  \n",
      " |  requires_gtf\n",
      " |  \n",
      " |  requires_protein_fasta\n",
      " |  \n",
      " |  requires_transcript_fasta\n",
      " |  \n",
      " |  transcript_sequences\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from serializable.serializable.Serializable:\n",
      " |  \n",
      " |  __reduce__(self)\n",
      " |      Overriding this method directs the default pickler to reconstruct\n",
      " |      this object using our from_dict method.\n",
      " |  \n",
      " |  to_json(self)\n",
      " |      Returns a string containing a JSON representation of this Genome.\n",
      " |  \n",
      " |  write_json_file(self, path)\n",
      " |      Serialize this VariantCollection to a JSON representation and write it\n",
      " |      out to a text file.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Class methods inherited from serializable.serializable.Serializable:\n",
      " |  \n",
      " |  _reconstruct_nested_objects(state_dict) from builtins.type\n",
      " |      Nested serializable objects will be represented as dictionaries so we\n",
      " |      allow manual reconstruction of those objects in this method.\n",
      " |      \n",
      " |      By default just returns the state dictionary unmodified.\n",
      " |  \n",
      " |  _update_kwargs(kwargs) from builtins.type\n",
      " |      Rename any old keyword arguments to preserve backwards compatibility\n",
      " |  \n",
      " |  from_json(json_string) from builtins.type\n",
      " |      Reconstruct an instance from a JSON string.\n",
      " |  \n",
      " |  read_json_file(path) from builtins.type\n",
      " |      Construct a VariantCollection from a JSON file.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors inherited from serializable.serializable.Serializable:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables (if defined)\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object (if defined)\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data and other attributes inherited from serializable.serializable.Serializable:\n",
      " |  \n",
      " |  _SERIALIZABLE_KEYWORD_ALIASES = {}\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#upload data, use pyensembl (pypi.org/project/pyensembl/)\n",
    "\n",
    "from pyensembl import EnsemblRelease\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "\n",
    "help(EnsemblRelease)\n",
    "\n",
    "\n",
    "#List all cDNA/cds sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SequenceData(fasta_paths=['/home/monica/.cache/pyensembl/GRCh38/ensembl99/Homo_sapiens.GRCh38.pep.all.fa.gz'])"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ENSP00000451515.1', 'ENSP00000451042.1', 'ENSP00000452494.1', 'ENSP00000488240.1', 'ENSP00000487941.1', 'ENSP00000419773.1', 'ENSP00000430034.1', 'ENSP00000488695.1', 'ENSP00000488000.1']\n",
      "111047\n",
      "[Seq('PSY', SingleLetterAlphabet()), Seq('EI', SingleLetterAlphabet()), Seq('TGGY', SingleLetterAlphabet()), Seq('GTGG', SingleLetterAlphabet()), Seq('GTGG', SingleLetterAlphabet()), Seq('VLLWFGELL', SingleLetterAlphabet()), Seq('*LQ*L', SingleLetterAlphabet()), Seq('LTG', SingleLetterAlphabet()), Seq('GIVGAT', SingleLetterAlphabet())]\n",
      "111047\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "protein_seqs = []\n",
    "protein_ids = []\n",
    "for record in SeqIO.parse(\"Data/Homo_sapiens.GRCh38.pep.all.fa\", \"fasta\"):\n",
    "    protein_seqs.append(record.seq)\n",
    "    protein_ids.append(record.id)\n",
    "\n",
    "print(protein_ids[0:9])\n",
    "print(len(protein_ids))\n",
    "print(protein_seqs[0:9])\n",
    "print(len(protein_seqs))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                       ID                                          Sequences\n",
      "0       ENSP00000451515.1                                          (P, S, Y)\n",
      "1       ENSP00000451042.1                                             (E, I)\n",
      "2       ENSP00000452494.1                                       (T, G, G, Y)\n",
      "3       ENSP00000488240.1                                       (G, T, G, G)\n",
      "4       ENSP00000487941.1                                       (G, T, G, G)\n",
      "...                   ...                                                ...\n",
      "111042  ENSP00000494625.1  (R, Q, G, R, C, D, T, Y, A, T, E, F, D, L, E, ...\n",
      "111043  ENSP00000494933.1  (M, A, G, R, R, V, N, V, N, V, G, V, L, G, H, ...\n",
      "111044  ENSP00000495578.1  (M, A, G, R, R, V, N, V, N, V, G, V, L, G, H, ...\n",
      "111045  ENSP00000496548.1  (M, P, S, M, L, E, R, I, S, K, N, L, V, K, E, ...\n",
      "111046  ENSP00000494855.1  (M, P, S, M, L, E, R, I, S, K, N, L, V, K, E, ...\n",
      "\n",
      "[111047 rows x 2 columns]\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "series_ids = pd.Series(protein_ids)\n",
    "series_seqs = pd.Series(protein_seqs)\n",
    "frame = {'ID': protein_ids, 'Sequences': protein_seqs}\n",
    "result = pd.DataFrame(frame)\n",
    "print(result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<_io.TextIOWrapper encoding='iso-8859-1'>"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. Gene transcription and translation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Transcribe & translate\n",
    "#Dictionaries: transcription and translation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3. List all proteins containing c-terminal degrons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Find c-degrons\n",
    "#Use regular expressions\n",
    "#List and enumerate all proteins containing c-degrons.\n",
    "#Which proportion of the total are?"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
